### PR TITLE
Make sure we Dispose() the options processor of CPS projects

### LINKED ISSUE #38443: Checking "Enable nullable references analysis IDE features" option crashes VS
**Version Used**: 
Version 16.3.0 Preview 3.0 [29230.54.d16.3]

**Steps to Reproduce**:
1. `git clone http://github.com/dotnet/project-system`
2. `cd project-system`
3. `git checkout 344a305681c5dc630f82f86880bc0a2e527fad1f`
4. `ProjectSystem.sln`
5. Tools -> Options -> Text Editor -> C# -> Advanced -> Enable nullable references analysis IDE features

```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.ArgumentException: '<Project852a3940-e0cd-4e7d-99e5-691190c256d9>' is not part of the workspace.
   at Microsoft.CodeAnalysis.Workspace.CheckProjectIsInCurrentSolution(ProjectId projectId)
   at Microsoft.CodeAnalysis.Workspace.HandleProjectChange(ProjectId projectId, Func`2 getSolutionWithChangedProject)
   at Microsoft.CodeAnalysis.Workspace.OnParseOptionsChanged(ProjectId projectId, ParseOptions options)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.<>c__DisplayClass49_0.<set_ParseOptions>b__1(Workspace w)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyChangeToWorkspace(Action`1 action)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.ChangeProjectProperty[T](T& field, T newValue, Func`2 withNewValue, Action`1 changeValue)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.set_ParseOptions(ParseOptions value)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectOptionsProcessor.UpdateProjectOptions_NoLock()
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectOptionsProcessor.UpdateProjectForNewHostValues()
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectOptionsProcessor.OptionService_OptionChanged(Object sender, OptionChangedEventArgs e)
   at Microsoft.CodeAnalysis.Options.OptionServiceFactory.OptionService.<>c__DisplayClass7_0.<OnGlobalOptionServiceOptionChanged>b__0()
   at Roslyn.Utilities.TaskExtensions.<>c__DisplayClass14_0.<ContinueWithAfterDelay>b__1(Task _)
   at Roslyn.Utilities.TaskExtensions.<>c__DisplayClass2_0.<SafeContinueWith>g__continuationFunction|0(Task antecedent)
   at Roslyn.Utilities.TaskExtensions.<>c__DisplayClass6_0`1.<SafeContinueWith>g__outerFunction|0(Task t)
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.ReportUnlessCanceled(System.Exception)
   at Roslyn.Utilities.TaskExtensions+<>c__DisplayClass6_0`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<SafeContinueWith>g__outerFunction|0(System.Threading.Tasks.Task)
   at Microsoft.CodeAnalysis.Workspace.CheckProjectIsInCurrentSolution(Microsoft.CodeAnalysis.ProjectId)
   at Microsoft.CodeAnalysis.Workspace.HandleProjectChange(Microsoft.CodeAnalysis.ProjectId, System.Func`2<Microsoft.CodeAnalysis.Solution,Microsoft.CodeAnalysis.Solution>)
   at Microsoft.CodeAnalysis.Workspace.OnParseOptionsChanged(Microsoft.CodeAnalysis.ProjectId, Microsoft.CodeAnalysis.ParseOptions)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject+<>c__DisplayClass49_0.<set_ParseOptions>b__1(Microsoft.CodeAnalysis.Workspace)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyChangeToWorkspace(System.Action`1<Microsoft.CodeAnalysis.Workspace>)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.ChangeProjectProperty[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.__Canon ByRef, System.__Canon, System.Func`2<Microsoft.CodeAnalysis.Solution,Microsoft.CodeAnalysis.Solution>, System.Action`1<Microsoft.CodeAnalysis.Workspace>)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.set_ParseOptions(Microsoft.CodeAnalysis.ParseOptions)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectOptionsProcessor.UpdateProjectOptions_NoLock()
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectOptionsProcessor.UpdateProjectForNewHostValues()
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectOptionsProcessor.OptionService_OptionChanged(System.Object, Microsoft.CodeAnalysis.Options.OptionChangedEventArgs)
   at Microsoft.CodeAnalysis.Options.OptionServiceFactory+OptionService+<>c__DisplayClass7_0.<OnGlobalOptionServiceOptionChanged>b__0()
   at Roslyn.Utilities.TaskExtensions+<>c__DisplayClass14_0.<ContinueWithAfterDelay>b__1(System.Threading.Tasks.Task)
   at Roslyn.Utilities.TaskExtensions+<>c__DisplayClass2_0.<SafeContinueWith>g__continuationFunction|0(System.Threading.Tasks.Task)
   at Roslyn.Utilities.TaskExtensions+<>c__DisplayClass6_0`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<SafeContinueWith>g__outerFunction|0(System.Threading.Tasks.Task)
   at System.Threading.Tasks.ContinuationResultTaskFromTask`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean)
   at System.Threading.Tasks.TaskScheduler.TryExecuteTask(System.Threading.Tasks.Task)
   at Microsoft.VisualStudio.LanguageServices.Implementation.VisualStudioTaskSchedulerFactory+JoinableTaskFactoryTaskScheduler+<>c__DisplayClass5_0+<<QueueTask>b__0>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory+SingleExecuteProtector.TryExecute()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory+SingleExecuteProtector+<>c.<.cctor>b__20_0(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
```

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Commit at a time recommended, as the first commit is just me adding nullable annotations.

- Fixes #38443
- Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/951410

### CODE REVIEW

File: src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs [hunk 1]
```diff
@@ -23,12 +25,12 @@ internal sealed partial class CPSProject : IWorkspaceProjectContext
...
         /// have the ability to parse command line options.
         /// </summary>
-        private readonly VisualStudioProjectOptionsProcessor _visualStudioProjectOptionsProcessor;
+        private readonly VisualStudioProjectOptionsProcessor? _visualStudioProjectOptionsProcessor;
 
         private readonly VisualStudioWorkspaceImpl _visualStudioWorkspace;
         private readonly IProjectCodeModel _projectCodeModel;
-        private readonly Lazy<ProjectExternalErrorReporter> _externalErrorReporterOpt;
-        private readonly EditAndContinue.VsENCRebuildableProjectImpl _editAndContinueProject;
+        private readonly Lazy<ProjectExternalErrorReporter?> _externalErrorReporterOpt;
```
Comment (sharwell) [inline src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs]:
```suggestion
        private readonly Lazy<ProjectExternalErrorReporter?> _externalErrorReporter;
```

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs]: I chose to skip this _for now_: this class is partial and there are uses in the other parts that aren't annotated yet, and I didn't want to annotate more than necessary if we're taking this through QB.

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs]: Actually maybe doing that one file won't be a problem either.

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs]: Fixed.

**Pull request merged**